qutebrowser/misc/notree.py

```diff
@@ -0,0 +1,355 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tree library for tree-tabs.
+
+The fundamental unit is the Node class.
+
+Create a tree with with Node(value, parent):
+root = Node('foo')
+child = Node('bar', root)
+child2 = Node('baz', root)
+child3 = Node('lorem', child)
+
+You can also assign parent after instantiation, or even reassign it:
+child4 = Node('ipsum')
+child4.parent = root
+
+Assign children:
+child.children = []
+child2.children = [child4, child3]
+child3.parent
+> Node('foo/bar/baz')
+
+Render a tree with render_tree(root_node):
+render_tree(root)
+
+> ('', 'foo')
+> ('├─', 'bar')
+> ('│ ├─', 'lorem')
+> ('│ └─', 'ipsum')
+> ('└─', 'baz')
+"""
+import enum
+import typing
+
+# For Node.render
+CORNER = '└─'
+INTERSECTION = '├─'
+PIPE = '│'
+
+
+class TreeError(RuntimeError):
+    """Exception used for tree-related errors."""
+
+
+class TraverseOrder(enum.Enum):
+    """To be used as argument to traverse().
+
+    Implemented orders are pre-order and post-order.
+    Attributes:
+        PRE: pre-order (parents before children). Same as in Node.render
+        POST: children of a node are always yield before their parent.
+        POST_R: Like POST, but children are yield in reverse order
+    """
+
+    PRE = 'pre-order'
+    POST = 'post-order'
+    POST_R = 'post-order-reverse'
+
+
+class Node():
+    """Fundamental unit of notree library.
+
+    Attributes:
+        value: The element (ususally a tab) the node represents
+        parent: Node's parent.
+        children: Node's children elements.
+        siblings: Children of parent node that are not self.
+        path: List of nodes from root of tree to self value, parent, and
+            children can all be set by user. Everything else will be updated
+            accordingly, so that if `node.parent = root_node`, then `node in
+            root_node.children` will be True.
+    """
+
+    sep = '/'  # type: str
+    __parent = None  # type: typing.Optional['Node']
+    # this is a global 'static' class attribute
+    __last_uid = 0  # type: int
+
+    PARENT_TYPE = typing.TypeVar('PARENT_TYPE')
+
+    def __init__(self,
+                 value: PARENT_TYPE,
+                 parent: typing.Optional['Node'] = None,
+                 childs: typing.Sequence['Node'] = (),
+                 uid: typing.Optional[int] = None) -> None:
+        if uid is not None:
+            self.__uid = uid  # TODO check if given ID exists already
+        else:
+            self.__uid = self.__get_new_uid()
```

Comment:
It turns out that the tab widget has a `tab_id` field which has a unique tab id (which always increases). It might be a little simpler to use that here instead, but I don't think it makes too much of a difference, so if it complicates things, I'm fine with leaving as is.

Reply:
Oh, I did not realize that there already was an uid. 

I guess you're right, it doesn't really make any difference, all it would change is two lines of code. I think I'll keep them distinct: otherwise, node uids and tab uids would be the same most of the time, with a few notable exceptions (undo'ed tabs, hidden and shown again...), which would be confusing.

However, I think I'll borrow the usage of itertools.count to generate numbers.